File: android/app/src/main/kotlin/com/myapp/cyber_guard/SecurePdfPlugin.kt
package com.myapp.cyber_guard

import android.content.Context
import android.graphics.Bitmap
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.nio.ByteBuffer
import java.util.UUID

/**
 * Secure PDF Plugin for Android.
 *
 * Uses Android's [PdfRenderer] API to render PDF pages as RGBA pixel data.
 * The Dart side receives raw pixels and displays them via [RawImage].
 *
 * ## MethodChannel: "com.cyberguard.security/pdf"
 *
 * ## Methods:
 * - open(source) → {documentId, pageCount}
 * - renderPage(documentId, pageIndex, width, height) → {pixels, width, height}
 * - getPageSize(documentId, pageIndex) → {width, height}
 * - searchText(documentId, query) → [{page, text}]  (basic — PdfRenderer has no text API)
 * - close(documentId)
 */
class SecurePdfPlugin : FlutterPlugin, MethodCallHandler {

    companion object {
        private const val TAG = "CyberGuard:PDF"
        private const val CHANNEL_NAME = "com.cyberguard.security/pdf"
    }

    private var channel: MethodChannel? = null
    private var context: Context? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    /** Active PDF documents (documentId → PdfDocument) */
    private val documents = mutableMapOf<String, PdfDocument>()

    // ─── FlutterPlugin ───

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME).also {
            it.setMethodCallHandler(this)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        context = null
        // Close all open documents
        documents.values.forEach { it.close() }
        documents.clear()
        scope.cancel()
    }

    // ─── MethodCallHandler ───

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "open" -> handleOpen(call, result)
            "renderPage" -> handleRenderPage(call, result)
            "getPageSize" -> handleGetPageSize(call, result)
            "searchText" -> handleSearchText(call, result)
            "close" -> handleClose(call, result)
            else -> result.notImplemented()
        }
    }

    // ─── Open ───

    private fun handleOpen(call: MethodCall, result: Result) {
        val source = call.argument<Map<String, Any>>("source")
        if (source == null) {
            result.error("INVALID_ARGS", "Missing source", null)
            return
        }

        scope.launch {
            try {
                val file = resolveSourceToFile(source)
                if (file == null) {
                    withContext(Dispatchers.Main) {
                        result.error("INVALID_SOURCE", "Could not resolve PDF source", null)
                    }
                    return@launch
                }

                val fd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
                val renderer = PdfRenderer(fd)
                val documentId = UUID.randomUUID().toString()

                documents[documentId] = PdfDocument(
                    id = documentId,
                    renderer = renderer,
                    fileDescriptor = fd,
                    tempFile = if (source["type"] == "network" || source["type"] == "memory") file else null
                )

                withContext(Dispatchers.Main) {
                    result.success(
                        mapOf(
                            "documentId" to documentId,
                            "pageCount" to renderer.pageCount,
                        )
                    )
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to open PDF: ${e.message}", e)
                withContext(Dispatchers.Main) {
                    result.error("OPEN_FAILED", e.message, null)
                }
            }
        }
    }

    // ─── Render Page ───

    private fun handleRenderPage(call: MethodCall, result: Result) {
        val documentId = call.argument<String>("documentId")
        val pageIndex = call.argument<Int>("pageIndex") ?: -1
        val width = call.argument<Int>("width") ?: 800
        val height = call.argument<Int>("height") ?: 1200

        val doc = documents[documentId]
        if (doc == null) {
            result.error("DOC_NOT_FOUND", "No document with id: $documentId", null)
            return
        }

        scope.launch {
            try {
                val page = doc.renderer.openPage(pageIndex)

                // Render to bitmap at requested dimensions
                val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
                page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
                page.close()

                // Convert ARGB_8888 to RGBA bytes for Flutter
                val buffer = ByteBuffer.allocate(bitmap.byteCount)
                bitmap.copyPixelsToBuffer(buffer)
                val argbBytes = buffer.array()
                val rgbaBytes = convertArgbToRgba(argbBytes)
                bitmap.recycle()

                withContext(Dispatchers.Main) {
                    result.success(
                        mapOf(
                            "pixels" to rgbaBytes,
                            "width" to width,
                            "height" to height,
                        )
                    )
                }
            } catch (e: Exception) {
                Log.e(TAG, "Render error page $pageIndex: ${e.message}", e)
                withContext(Dispatchers.Main) {
                    result.error("RENDER_FAILED", e.message, null)
                }
            }
        }
    }

    // ─── Get Page Size ───

    private fun handleGetPageSize(call: MethodCall, result: Result) {
        val documentId = call.argument<String>("documentId")
        val pageIndex = call.argument<Int>("pageIndex") ?: 0

        val doc = documents[documentId]
        if (doc == null) {
            result.error("DOC_NOT_FOUND", "No document with id: $documentId", null)
            return
        }

        try {
            val page = doc.renderer.openPage(pageIndex)
            val width = page.width.toDouble()
            val height = page.height.toDouble()
            page.close()

            result.success(
                mapOf(
                    "width" to width,
                    "height" to height,
                )
            )
        } catch (e: Exception) {
            result.error("PAGE_SIZE_FAILED", e.message, null)
        }
    }

    // ─── Search Text ───

    private fun handleSearchText(call: MethodCall, result: Result) {
        // Android PdfRenderer does not support text extraction.
        // Return empty results — text search requires a more advanced library.
        result.success(emptyList<Map<String, Any>>())
    }

    // ─── Close ───

    private fun handleClose(call: MethodCall, result: Result) {
        val documentId = call.argument<String>("documentId")
        val doc = documents.remove(documentId)
        doc?.close()
        result.success(null)
    }

    // ─── Source Resolution ───

    private suspend fun resolveSourceToFile(source: Map<String, Any>): File? {
        val type = source["type"] as? String ?: "network"
        val ctx = context ?: return null

        return when (type) {
            "network" -> {
                val url = source["url"] as? String ?: return null
                @Suppress("UNCHECKED_CAST")
                val headers = source["headers"] as? Map<String, String> ?: emptyMap()
                downloadToTemp(ctx, url, headers)
            }
            "asset" -> {
                val assetPath = source["assetPath"] as? String ?: return null
                copyAssetToTemp(ctx, assetPath)
            }
            "file" -> {
                val filePath = source["filePath"] as? String ?: return null
                File(filePath).takeIf { it.exists() }
            }
            "memory" -> {
                val bytes = source["bytes"] as? ByteArray ?: return null
                writeBytesToTemp(ctx, bytes)
            }
            else -> null
        }
    }

    private suspend fun downloadToTemp(ctx: Context, urlStr: String, headers: Map<String, String>): File? {
        return withContext(Dispatchers.IO) {
            try {
                val url = URL(urlStr)
                val conn = url.openConnection() as HttpURLConnection
                headers.forEach { (key, value) -> conn.setRequestProperty(key, value) }
                conn.connectTimeout = 30_000
                conn.readTimeout = 30_000

                val tempFile = File.createTempFile("cyberguard_pdf_", ".pdf", ctx.cacheDir)
                conn.inputStream.use { input ->
                    FileOutputStream(tempFile).use { output ->
                        input.copyTo(output)
                    }
                }
                conn.disconnect()
                tempFile
            } catch (e: Exception) {
                Log.e(TAG, "Download failed: ${e.message}", e)
                null
            }
        }
    }

    private fun copyAssetToTemp(ctx: Context, assetPath: String): File? {
        return try {
            val key = ctx.assets.openFd("flutter_assets/$assetPath")
            val tempFile = File.createTempFile("cyberguard_pdf_", ".pdf", ctx.cacheDir)
            key.createInputStream().use { input ->
                FileOutputStream(tempFile).use { output ->
                    input.copyTo(output)
                }
            }
            tempFile
        } catch (e: Exception) {
            Log.e(TAG, "Asset copy failed: ${e.message}", e)
            null
        }
    }

    private fun writeBytesToTemp(ctx: Context, bytes: ByteArray): File? {
        return try {
            val tempFile = File.createTempFile("cyberguard_pdf_", ".pdf", ctx.cacheDir)
            FileOutputStream(tempFile).use { it.write(bytes) }
            tempFile
        } catch (e: Exception) {
            Log.e(TAG, "Memory write failed: ${e.message}", e)
            null
        }
    }

    // ─── Pixel Conversion ───

    /**
     * Convert Android's ARGB_8888 to RGBA_8888 expected by Flutter.
     *
     * ARGB: [A, R, G, B] per pixel (Android native)
     * RGBA: [R, G, B, A] per pixel (Flutter ui.PixelFormat.rgba8888)
     */
    private fun convertArgbToRgba(argb: ByteArray): ByteArray {
        val rgba = ByteArray(argb.size)
        var i = 0
        while (i < argb.size) {
            rgba[i] = argb[i + 1]     // R
            rgba[i + 1] = argb[i + 2] // G
            rgba[i + 2] = argb[i + 3] // B
            rgba[i + 3] = argb[i]     // A
            i += 4
        }
        return rgba
    }
}

/**
 * Holds an open PdfRenderer and its associated resources.
 */
private data class PdfDocument(
    val id: String,
    val renderer: PdfRenderer,
    val fileDescriptor: ParcelFileDescriptor,
    val tempFile: File?,
) {
    fun close() {
        try { renderer.close() } catch (_: Exception) {}
        try { fileDescriptor.close() } catch (_: Exception) {}
        tempFile?.delete()
    }
}

File: ios/Runner/Security/SecurePdfPlugin.swift
import Flutter
import UIKit
import PDFKit

/// Secure PDF Plugin for iOS.
///
/// Uses PDFKit (`PDFDocument`, `PDFPage`) to render PDF pages as RGBA pixel data.
/// The Dart side receives raw pixels and displays them via `RawImage`.
///
/// ## Channel: "com.cyberguard.security/pdf"
///
/// ## Methods:
/// - open(source) → {documentId, pageCount}
/// - renderPage(documentId, pageIndex, width, height) → {pixels, width, height}
/// - getPageSize(documentId, pageIndex) → {width, height}
/// - searchText(documentId, query) → [{page, text}]
/// - close(documentId)
class SecurePdfPlugin: NSObject, FlutterPlugin {

    static let channelName = "com.cyberguard.security/pdf"

    private var channel: FlutterMethodChannel?

    /// Active PDF documents (documentId → PdfDocumentHandle)
    private var documents: [String: PdfDocumentHandle] = [:]

    // MARK: - FlutterPlugin

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = SecurePdfPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "open":
            handleOpen(call, result: result)
        case "renderPage":
            handleRenderPage(call, result: result)
        case "getPageSize":
            handleGetPageSize(call, result: result)
        case "searchText":
            handleSearchText(call, result: result)
        case "close":
            handleClose(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Open

    private func handleOpen(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let sourceMap = args["source"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing source", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let pdfDocument = self?.resolvePdfDocument(from: sourceMap) else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "OPEN_FAILED", message: "Could not open PDF", details: nil))
                }
                return
            }

            let documentId = UUID().uuidString
            let handle = PdfDocumentHandle(document: pdfDocument)

            DispatchQueue.main.async {
                self?.documents[documentId] = handle
                result([
                    "documentId": documentId,
                    "pageCount": pdfDocument.pageCount,
                ] as [String: Any])
            }
        }
    }

    // MARK: - Render Page

    private func handleRenderPage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let documentId = args["documentId"] as? String,
              let pageIndex = args["pageIndex"] as? Int,
              let width = args["width"] as? Int,
              let height = args["height"] as? Int else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing required parameters", details: nil))
            return
        }

        guard let handle = documents[documentId],
              let page = handle.document.page(at: pageIndex) else {
            result(FlutterError(code: "DOC_NOT_FOUND", message: "Document or page not found", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            let pageBounds = page.bounds(for: .mediaBox)

            // Create RGBA bitmap context
            let colorSpace = CGColorSpace(name: CGColorSpace.sRGB)!
            guard let ctx = CGContext(
                data: nil,
                width: width,
                height: height,
                bitsPerComponent: 8,
                bytesPerRow: width * 4,
                space: colorSpace,
                bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue // RGBA
            ) else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "RENDER_FAILED", message: "Could not create bitmap context", details: nil))
                }
                return
            }

            // White background
            ctx.setFillColor(UIColor.white.cgColor)
            ctx.fill(CGRect(x: 0, y: 0, width: width, height: height))

            // Scale PDF page to fit
            let scaleX = CGFloat(width) / pageBounds.width
            let scaleY = CGFloat(height) / pageBounds.height
            let scale = min(scaleX, scaleY)

            let offsetX = (CGFloat(width) - pageBounds.width * scale) / 2
            let offsetY = (CGFloat(height) - pageBounds.height * scale) / 2

            ctx.translateBy(x: offsetX, y: offsetY)
            ctx.scaleBy(x: scale, y: scale)

            // PDFPage draws with origin at bottom-left; CGContext is also bottom-left
            if let pageRef = page.pageRef {
                ctx.drawPDFPage(pageRef)
            }

            // Extract pixel data
            guard let data = ctx.data else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "RENDER_FAILED", message: "No pixel data", details: nil))
                }
                return
            }

            let byteCount = width * height * 4
            let pixels = Data(bytes: data, count: byteCount)

            DispatchQueue.main.async {
                result([
                    "pixels": FlutterStandardTypedData(bytes: pixels),
                    "width": width,
                    "height": height,
                ] as [String: Any])
            }
        }
    }

    // MARK: - Get Page Size

    private func handleGetPageSize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let documentId = args["documentId"] as? String,
              let pageIndex = args["pageIndex"] as? Int else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing parameters", details: nil))
            return
        }

        guard let handle = documents[documentId],
              let page = handle.document.page(at: pageIndex) else {
            result(FlutterError(code: "DOC_NOT_FOUND", message: "Document or page not found", details: nil))
            return
        }

        let bounds = page.bounds(for: .mediaBox)
        result([
            "width": Double(bounds.width),
            "height": Double(bounds.height),
        ] as [String: Any])
    }

    // MARK: - Search Text

    private func handleSearchText(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let documentId = args["documentId"] as? String,
              let query = args["query"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing parameters", details: nil))
            return
        }

        guard let handle = documents[documentId] else {
            result(FlutterError(code: "DOC_NOT_FOUND", message: "Document not found", details: nil))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            let selections = handle.document.findString(query, withOptions: .caseInsensitive)
            var results: [[String: Any]] = []

            for selection in selections {
                if let page = selection.pages.first {
                    let pageIndex = handle.document.index(for: page)
                    results.append([
                        "page": pageIndex,
                        "text": selection.string ?? query,
                    ])
                }
            }

            DispatchQueue.main.async {
                result(results)
            }
        }
    }

    // MARK: - Close

    private func handleClose(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let documentId = args["documentId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing documentId", details: nil))
            return
        }

        documents.removeValue(forKey: documentId)
        result(nil)
    }

    // MARK: - Source Resolution

    private func resolvePdfDocument(from source: [String: Any]) -> PDFDocument? {
        let type = source["type"] as? String ?? "network"

        switch type {
        case "network":
            guard let urlStr = source["url"] as? String,
                  let url = URL(string: urlStr) else { return nil }
            // Synchronous download (called on background thread)
            guard let data = try? Data(contentsOf: url) else { return nil }
            return PDFDocument(data: data)

        case "asset":
            guard let assetPath = source["assetPath"] as? String else { return nil }
            let key = FlutterDartProject.lookupKey(forAsset: assetPath)
            guard let path = Bundle.main.path(forResource: key, ofType: nil) else { return nil }
            guard let url = URL(string: "file://\(path)") else { return nil }
            return PDFDocument(url: url)

        case "file":
            guard let filePath = source["filePath"] as? String else { return nil }
            let url = URL(fileURLWithPath: filePath)
            return PDFDocument(url: url)

        case "memory":
            guard let bytes = source["bytes"] as? FlutterStandardTypedData else { return nil }
            return PDFDocument(data: bytes.data)

        default:
            return nil
        }
    }
}

/// Holds an open PDFDocument reference.
private class PdfDocumentHandle {
    let document: PDFDocument

    init(document: PDFDocument) {
        self.document = document
    }
}
